subxt/src/backend/legacy/mod.rs

```diff
@@ -84,183 +85,249 @@ impl<T: Config + Send + Sync + 'static> Backend<T> for LegacyBackend<T> {
-        let Some(details) = self.methods.chain_get_block(Some(at)).await? else {
-            return Ok(None);
-        };
-        Ok(Some(
-            details.block.extrinsics.into_iter().map(|b| b.0).collect(),
-        ))
+        retry(|| async {
+            let Some(details) = self.methods.chain_get_block(Some(at)).await? else {
+                return Ok(None);
+            };
+            Ok(Some(
+                details.block.extrinsics.into_iter().map(|b| b.0).collect(),
+            ))
+        })
+        .await
     }
 
     async fn latest_finalized_block_ref(&self) -> Result<BlockRef<T::Hash>, Error> {
-        let hash = self.methods.chain_get_finalized_head().await?;
-        Ok(BlockRef::from_hash(hash))
+        retry(|| async {
+            let hash = self.methods.chain_get_finalized_head().await?;
+            Ok(BlockRef::from_hash(hash))
+        })
+        .await
     }
 
     async fn current_runtime_version(&self) -> Result<RuntimeVersion, Error> {
-        let details = self.methods.state_get_runtime_version(None).await?;
-        Ok(RuntimeVersion {
-            spec_version: details.spec_version,
-            transaction_version: details.transaction_version,
+        retry(|| async {
+            let details = self.methods.state_get_runtime_version(None).await?;
+            Ok(RuntimeVersion {
+                spec_version: details.spec_version,
+                transaction_version: details.transaction_version,
+            })
         })
+        .await
     }
 
     async fn stream_runtime_version(&self) -> Result<StreamOfResults<RuntimeVersion>, Error> {
-        let sub = self.methods.state_subscribe_runtime_version().await?;
-        let sub = sub.map(|r| {
-            r.map(|v| RuntimeVersion {
-                spec_version: v.spec_version,
-                transaction_version: v.transaction_version,
+        let methods = self.methods.clone();
+
+        let retry_sub = retry_stream(move || {
+            let methods = methods.clone();
+
+            Box::pin(async move {
+                let sub = methods.state_subscribe_runtime_version().await?;
+                let sub = sub.map(|r| {
+                    r.map(|v| RuntimeVersion {
+                        spec_version: v.spec_version,
+                        transaction_version: v.transaction_version,
+                    })
+                });
+                Ok(StreamOf(Box::pin(sub)))
             })
-        });
-        Ok(StreamOf(Box::pin(sub)))
+        })
+        .await?;
+
+        Ok(retry_sub)
     }
 
     async fn stream_all_block_headers(
         &self,
     ) -> Result<StreamOfResults<(T::Header, BlockRef<T::Hash>)>, Error> {
-        let sub = self.methods.chain_subscribe_all_heads().await?;
-        let sub = sub.map(|r| {
-            r.map(|h| {
-                let hash = h.hash();
-                (h, BlockRef::from_hash(hash))
+        let methods = self.methods.clone();
+
+        let retry_sub = retry_stream(move || {
```

Comment:
`stream_all_block_headers`, `stream_best_block_headers` and `stream_finalized_block_headers` are a bit more complicated I think.

At the moment, you're silently restarting the stream on reconnection. This possibly misses blocks as a result but the user would have no idea that they might have missed anything.

I think there are two options that prob make more sense:
1. Restart the stream on reconnection, but tell the user about the disconnection (ie pass on the `Err`) so that they know blocks may have been missed.
   - Means that the user has to explicitly accept that a disconnect happened (ie handle the error without bailing), else they will bail.
2. "catch up" and download any blocks missed, so that for the user it appears mostly seamless, and there are no important gaps.
   - In reality, we'll miss some best/new blocks (and blocks may get pruned etc while we are disconnected), but we can fill in any gaps by giving the finalized blocks.
   - `stream_finalized_block_headers` already uses `subscribe_to_block_headers_filling_in_gaps` which does something like this, so maybe we can make use of that, or similar,
     if need be.

It's probably easiest to do 1 to begin with, but being able to be seamless by doing 2 would be cool (and similar to what we'd want to do in the unstable backend).

Both here and in the unstable backend, if we aim to do 2 and can't be seamless (ie the blocks we need to fill in the gaps are no longer available to download) then we should emit an error and end.

Reply:
The same applies for runtime upgrades as well but maybe not as likely to occur.

2) that seems quite complicated and prefer not implement such a thing in this PR.

So I guess you are suggesting that we should emit a message or error message "MissedMessage(Vec<BlockRef>)" on the stream and it's to the user to decide whether to fetch missed the blocks?

Reply:
I have modified the RetryStream impl to emit `DisconnectWillReconnect` now, so should fine now